tests/identities/TestProvider.ts

```diff
@@ -82,6 +82,28 @@ class TestProvider extends Provider {
     };
   }
 
+  public resetLinks() {
```

Comment:
What does this function do?

Reply:
It removes all claims from the test provider, since these were affecting later tests. Addressed in this comment https://github.com/MatrixAI/js-polykey/pull/278#issuecomment-985230957

Reply:
It's not part of the original provider standard, and it seems special cased. This seems like something that should be done in the test function itself, not embedded in `TestProvider`.

Reply:
Looking back on this function again it's actually resetting the cryptolinks back to the default created during contruction rather than deleting all of them, so there still needs to be some way of keeping track of this default. There's also no existing way of directly accessing or editing the links stored on the test provider from outside of it, since the field is protected and there are no getters or setters, so this would need to change. I think keeping it the way it is is the simplest option, especially considering that the test provider is implemented very differently from a normal provider anyway and already has an extra function `overrideLinks()` that isn't part of the original provider standard and is purely there to make testing easier.

Reply:
So can this be done in the tests themselves? It's not really part of the domain model of `TestProvider` right?

Reply:
I didn't add `overrideLinks` into this class.

It was added in here: https://github.com/MatrixAI/js-polykey/commit/c2dfe01c281fef8adcee8de17263363c733ea800.

I don't think that had a full deep review, and I suspect that this is the wrong place to do it.

If you must, you can also make the links public so you can mutate them in the tests.

Reply:
Could you also just do a `ts-ignore` to access and mutate the links on a test-by-test basis? Or bad practice?

Reply:
Like the concept of `links` itself is specific to `TestProvider`. So you can just make that public if you need to modify those links.

I would also suggest just creating a new `IdentityProvider` instance when you want to change a "default state"... but I guess during testing this can be cost prohibitive.

Reply:
This is what I wrote originally:

```ts
  protected linkIdCounter: number = 0;
  protected users: Record<IdentityId, POJO>;
  protected links: Record<LinkId, string>;
  protected userLinks: Record<IdentityId, Array<LinkId>>;
  protected userTokens: Record<string, IdentityId>;

  public constructor() {
    super();
    this.users = {
      test_user: {
        email: 'test_user@test.com',
      },
      test_user2: {
        email: 'test_user2@test.com',
      },
    };
    this.links = {
      test_link: JSON.stringify({
        type: 'identity',
        node: 'nodeid' as NodeId,
        provider: this.id,
        identity: 'test_link' as IdentityId,
        timestamp: 1618650105,
        signature: 'somesignature',
      }),
    };
    this.userLinks = {
      test_user: ['test_link'],
    };
    this.userTokens = {
      abc123: 'test_user' as IdentityId,
    };
  }
```

The `links` property was only there to enable the `TestProvider` to seed it with some initial state.

If you want to do this properly, just make the `links` property public, and mutate it as you are testing it.

You could even change the `constructor` of `TestProvider` so that tests can seed the relevant state that you want.

But arbitrary methods inside the `TestProvider` is not the right place to do link state mutation. They don't correspond to any test, and it's difficult to know why these methods exist at all when looking at this code 6 months later.

Reply:
Ok, I'll remove the two link-mutating metods and make the `links` property public.

Reply:
This has been fixed